### PR TITLE
crimson: simplify snaptrim operation pipline usage

### PR DESCRIPTION
**Pull request by athanatos**
<details>
<summary>Show available Jenkins commands</summary>

- `jenkins retest this please`
- `jenkins test classic perf`
- `jenkins test crimson perf`
- `jenkins test signed`
- `jenkins test make check`
- `jenkins test make check arm64`
- `jenkins test submodules`
- `jenkins test dashboard`
- `jenkins test dashboard cephadm`
- `jenkins test api`
- `jenkins test docs`
- `jenkins render docs`
- `jenkins test ceph-volume all`
- `jenkins test ceph-volume tox`
- `jenkins test windows`
- `jenkins test rook e2e`
</details>

### CODE REVIEW

File: src/crimson/osd/osd_operations/snaptrim_event.h [hunk 1]
```diff
@@ -153,12 +138,6 @@ class SnapTrimObjSubEvent : public PhasedOperationT<SnapTrimObjSubEvent> {
   remove_or_update_iertr::future<ceph::os::Transaction>
   remove_or_update(ObjectContextRef obc, ObjectContextRef head_obc);
 
-  // we don't need to synchronize with other instances started by
-  // SnapTrimEvent; it's here for the sake of op tracking.
```
Comment (Matan-B) [inline src/crimson/osd/osd_operations/snaptrim_event.h:158]: Should we reference a follow up tracker here to re-implement stages op tracking / event flags?

File: src/crimson/osd/osd_operations/snaptrim_event.cc [hunk 2]
```diff
@@ -414,65 +414,67 @@ SnapTrimObjSubEvent::remove_or_update(
 SnapTrimObjSubEvent::snap_trim_obj_subevent_ret_t
 SnapTrimObjSubEvent::start()
 {
-  return enter_stage<interruptor>(
+  auto exit_handle = seastar::defer([this] {
```
Comment (Matan-B) [inline src/crimson/osd/osd_operations/snaptrim_event.cc:384]: nit:
* [[maybe_unused]]
* `// Called on exit_handle destruction`

**Review by Matan-B — APPROVED** [review]
LGTM, few comments

File: src/crimson/osd/osd_operations/snaptrim_event.cc (hunk 2 above)
Comment (athanatos) [reply src/crimson/osd/osd_operations/snaptrim_event.cc:384]: I don't think it actually needs the [[maybe_unused]] -- the constructor/destructor have side effects.  "Called on exit_handle's destruction" merely explains what seastar::defer does -- that will mostly be unnecessary visual noise for anyone familiar with the code base and fairly easy to pick up from context for anyone with a basic understanding of C++ RAII.

File: src/crimson/osd/osd_operations/snaptrim_event.h (hunk 1 above)
Comment (athanatos) [reply src/crimson/osd/osd_operations/snaptrim_event.h:158]: Added a comment.

**Comment by Matan-B** [comment]
https://pulpito.ceph.com/matan-2024-06-17_15:38:16-crimson-rados-wip-matanb-crimson-testing-june-16-distro-crimson-smithi/

**Comment by Matan-B** [comment]
jenkins test api

**Comment by athanatos** [comment]
https://pulpito.ceph.com/sjust-2024-06-14_20:06:29-crimson-rados:thrash-wip-sjust-crimson-testing-2024-06-14-distro-default-smithi/

  - 7757004 - AsyncReserver https://tracker.ceph.com/issues/65696 (see above, removed https://github.com/ceph/ceph/pull/57147)
  - 7756998 - osd crash on shutdown, heap use-after-free on osdmap
    - https://tracker.ceph.com/issues/66478
  - 7757008 - load_pgs https://tracker.ceph.com/issues/66294
  - 7757016 - load_pgs https://tracker.ceph.com/issues/66294
  - 7757018 - AsyncReserver https://tracker.ceph.com/issues/66316

**Comment by athanatos** [comment]
I think this is ready to merged based on my results above.

**Comment by Matan-B** [comment]
> I think this is ready to merged based on my results above.

Yes, teseted with the stable `simple` thrashers:

https://pulpito.ceph.com/matan-2024-06-18_06:55:42-crimson-rados-wip-sjust-crimson-testing-2024-06-14-distro-crimson-smithi/

**Comment by Matan-B** [comment]
jenkins test api

**Pull request merged**